asynciterator.ts

```diff
@@ -1251,6 +1259,244 @@ function destinationFillBuffer<S>(this: InternalSource<S>) {
     (this._destination as any)._fillBuffer();
 }
 
+export class SynchronousTransformIterator<S, D = S> extends AsyncIterator<D> {
+  protected readonly _source: AsyncIterator<S>;
+
+  constructor(source: AsyncIterator<S>) {
+    /* eslint-disable no-use-before-define */
+    super();
+    this._source = source;
+    const cleanup = () => {
+      source.removeListener('end', onEnd);
+      source.removeListener('readable', onReadable);
+    };
+    const onEnd = () => {
+      cleanup();
+      this.close();
+    };
+    const onReadable = () => {
+      this.readable = true;
+    };
+    source.on('end', onEnd);
+    source.on('readable', onReadable);
+  }
+
+  protected _destroy(cause: Error | undefined, callback: (error?: Error) => void) {
+    super._destroy(cause, callback);
+    this._source.destroy(cause);
+  }
+}
+
+export class MultiMappingIterator<S, D = S> extends SynchronousTransformIterator<S, D> {
+  protected readonly _map: (item: S) => Generator<D>;
+  private generator?: Generator<D>;
+
+  constructor(source: AsyncIterator<S>, map: (item: S) => Generator<D>) {
+    super(source);
+    this._map = map;
+  }
+
+  read(): D | null {
+    let _item;
+
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+      if (!this.generator) {
+        if ((_item = this._source.read()) === null)
+          return null;
+        this.generator = this._map(_item);
+      }
+      if (!(_item = this.generator.next()).done)
+        return _item.value;
+      this.generator = undefined;
+    }
+  }
+}
+
+export class MappingIterator<S, D = S> extends SynchronousTransformIterator<S, D> {
+  protected readonly _map: (item: S) => D;
+
+  constructor(source: AsyncIterator<S>, map: (item: S) => D) {
+    super(source);
+    this._map = map;
+  }
+
+  read(): D | null {
+    const item = this._source.read();
+    if (item !== null)
+      return this._map(item);
+    return null;
+  }
+
+  map<T>(map: (item: D) => T, self?: any): AsyncIterator<T> {
+    return new MultiMapFilterTransformIterator(this._source, {
+      filter: false,
+      function: self ? map.bind(self) : map,
+      next: {
+        filter: false,
+        function: this._map,
+      },
+    });
+  }
+
+  filter(filter: (item: D) => boolean, self?: any): AsyncIterator<D> {
+    return new MultiMapFilterTransformIterator(this._source, {
+      filter: true,
+      function: self ? filter.bind(self) : filter,
+      next: {
+        filter: false,
+        function: this._map,
+      },
+    });
+  }
```

Comment:
Per [this comment](https://github.com/RubenVerborgh/AsyncIterator/pull/53#issuecomment-1079812777) it still may be worth removing these lines. @jacoscaz it might be worth doing some perf. tests on your machine to see if your results align with mine.

Reply:
I think the cause of the marginal slowdown observed there was the fact that we are essentially adding an extra function for every map operation in order to compose them `(item: any) => t(_func(item));`

Reply:
Curiously enough, I am getting 10% - 15% _faster_ iteration by keeping those lines in. Removing them seems to slow things down. On some level this makes sense as V8 should optimize execution of the "compiled" transformation more effectively. Another differentiating factor might be the number of item vs. the number of iterators. These lines likely produce higher setup latencies, so they might have a more significant impact on scenarios with low numbers of items and high numbers of iterators. I would keep these in and postpone further tuning to a later stage as we're currently lacking dedicated scaffolding for this sort of testing.